### DISCUSSION TITLE
Server API Route, no headers from client

### DISCUSSION THREAD

**Discussion by ghost** [nuxt/nuxt#16065] [Questions] [answered]
Hello, I am happy to participate in the beta of Nuxt 3 !

You should know that Nuxt 3 manages its own API server in the /server/api/(route.js) folder

However, this feature seems to be limited because we don't have access to the headers of the client request because it's empty

What would be interesting is to propagate the client headers through the server-side (SSR) in order to retrieve them in "req.headers".

Why should we do this?
Because when we develop an API, we need to have its cookies, to have the Accept-Language, the header host in case of a multi-domain application and some custom headers sent by the reverse-proxy (nginx for example)

What do you think about it? Have you found an alternative solution to manage this?

**Comment by MarcTroll** [comment]
In addition to this: There should also be the possibility to set cookies from inside API Routes, not just reading them from the client.

**Reply by wobsoriano** [reply]
You actually can :)

```ts
import type { IncomingMessage, ServerResponse } from 'http'
import { useQuery, setCookie } from 'h3'

export default async (req: IncomingMessage, res: ServerResponse) => {
    // Read
    const cookies = useQuery(req)

    // Set
    setCookie(res, 'name', 'value')

    return {
        success: true
    }
}
```

https://github.com/unjs/h3

**Reply by MarcTroll** [reply]
This doesn't work for me. Doing the following neither writes nor reads any cookies, although they're set.
```ts
import type { IncomingMessage, ServerResponse } from "http";
import {
	useCookies as getCookies,
	setCookie
} from 'h3';

export default async (req: IncomingMessage, res: ServerResponse) => {
	console.log(getCookies(req));
	setCookie(res, "userID", "1", {
		httpOnly: true
	});
	
	return {
		title: "Hello"
	}
}
```

When calling the API-routes directly (and not in SSR-context in nuxt using `useFetch`) this works as expected.

**Reply by eMeRiKa13** [reply]
@ilouHD I have the same problem. I'm trying to get a cookie in a SSR-context but getCookies returns nothing.

Maybe it's due to that:
> Nitro allows 'direct' calling of routes via the globally-available $fetch helper. This will make an API call to the server if run on the browser, but will simply call the relevant function if run on the server, saving an additional API call.

From https://dev.to/ajcwebdev/a-first-look-at-nuxt-3-1769

**Comment by P4sca1** [comment]
The same applies to plugins: With nuxt 2 it was possible to access the client request using `ctx.req`. This allowed to extract client headers for authentication (cookies) during SSR.

**Reply by wobsoriano** [reply]
```html
<script setup lang="ts">
const nuxtApp = useNuxtApp()

nuxtApp.ssrContext.req
</script>
```

**Reply by P4sca1** [reply]
Didn’t know that was possible. Types are missing that piece of Information: https://github.com/nuxt/framework/blob/035e9b5e8b5641fa778df131a4776ec584c7eddf/packages/nuxt3/src/app/nuxt.ts#L40

**Reply by wobsoriano** [reply]
Yeah I actually just saw it in their example https://v3.nuxtjs.org/docs/usage/state#example

**Comment by wobsoriano** [comment]
Nuxt uses [h3](https://github.com/unjs/h3) as its http framework. You can check the docs on how to access cookies, body, query params, etc.

```ts
import type { IncomingMessage, ServerResponse } from 'http'
import {
    useCookies,
    useQuery,
    useBody,
    setCookie,
    appendHeader
} from 'h3'

export default async (req: IncomingMessage, res: ServerResponse) => {
    // Read cookies
    const cookies = useCookies(req)

    // Set a cookie
    setCookie(res, 'name', 'value')

    // Read query params
    const query = useQuery(req)

    // Read body from post
    const body = useBody(req)

   // Append header
    appendHeader(res, 'name', 'value')

    return {
        success: true
    }
}
```

**Reply by mrdannyjohnson** [reply]
@wobsoriano thank you for the above code.

For me, using h3 useCookie() works perfectly when navigating around my small test app, but fails to send cookie headers when I reload a page. Not sure if this is something to do with how I'm setting the cookies, or a more general SSR problem. 

Have you experienced this? Or do you have any suggestions? Thank you.

**Reply by eMeRiKa13** [reply]
@mrdannyjohnson I have the same problem. When reloading the page, I'm trying to get a cookie in a SSR-context but getCookies returns nothing. Did you find a solution at this problem?

**Reply by cdwmhcc** [reply]
same~

**Reply by eMeRiKa13** [reply]
@cdwmhcc how do you call this route?

I had to manually set the cookie to make it work from a Pinia store:
```js
const user = await $fetch('/api/auth/get-user', {
  method: 'POST',
  headers: {
    cookie: this.$nuxt.ssrContext.req.headers['cookie']
  }
});
```

**Comment by cdwmhcc** [comment]
> I had to manually set the cookie to make it work from a Pinia store

Oh, I'm talking about the server, not the client.

app.vue

``` typescript
const { data: test } = await useFetch(`/api/test`)
```
~/server/api/test.ts

``` typescript
import type { IncomingMessage, ServerResponse } from 'http'
import { useCookies } from 'h3'
export default async (req: IncomingMessage, res: ServerResponse) => {
    const cookies = useCookies(req)
    console.log('headers', req.headers)
    console.log('cookies', cookies)
    return {}
}
```

SSR, it print:

```
headers { host: undefined }
cookies {}
```

Client, it work.

Hope that the headers of the ssr request can be passed for authentication.

@eMeRiKa13

**Reply by lewebsimple** [reply]
According to [this](http://v3.nuxtjs.org/docs/usage/cookies), useCookies will only work from the setup function or inside a lifecycle hook. So only h3 methods would work inside server api.

I highly recommend Atinux’ talk from VueConf Toronto 2021 for a simple authentication example (repository available [here](https://github.com/Atinux/discuss) )

**Comment by danielroe** [comment, marked as answer]
If you need to to access headers on a server API route + a local call, consider using `useReqHeaders` to proxy the headers to the local call. See [docs](https://v3.nuxtjs.org/docs/usage/data-fetching#isomorphic-fetch).

**Reply by Hyperblaster** [reply]
That link is now a 404. And after upgrading to the production version of nuxt, it's really hard to figure out all the things that have now broken
